certgen/certgen_test.go

```diff
@@ -0,0 +1,242 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certgen
+
+import (
+	"bytes"
+	"crypto"
+	"fmt"
+	"io/ioutil"
+	"testing"
+	"time"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509util"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/monologue/interval"
+	"github.com/google/trillian/crypto/keys/pem"
+)
+
+const (
+	rootFile    = "./testdata/test_ca.pem"
+	rootKeyFile = "./testdata/test_ca.key"
+
+	subjectCommonName         = "test-leaf-certificate"
+	subjectOrganization       = "Test Organisation"
+	subjectOrganizationalUnit = "Test Organisational Unit"
+	subjectLocality           = "Test Locality"
+	subjectCountry            = "GB"
+	signatureAlgorithm        = x509.SHA256WithRSA
+	prefix                    = "test-log"
+)
+
+func rootAndKeySetup(rootFile, rootKeyFile string) (*x509.Certificate, crypto.Signer, error) {
+	rootPEM, err := ioutil.ReadFile(rootFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error reading root cert: %s", err)
+	}
+	root, err := x509util.CertificateFromPEM(rootPEM)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing root cert: %s", err)
+	}
+
+	rootKeyPEM, err := ioutil.ReadFile(rootKeyFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error reading root key: %s", err)
+	}
+	rootKey, err := pem.UnmarshalPrivateKey(string(rootKeyPEM), "")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing root key: %s", err)
+	}
+
+	return root, rootKey, nil
+}
+
+func TestIssueCertificate(t *testing.T) {
+	tests := []struct {
+		desc             string
+		notAfterInterval *interval.Interval
+	}{
+		{
+			desc: "not temporal",
+		},
+		{
+			desc: "smallest temporal",
+			notAfterInterval: &interval.Interval{
+				Start: time.Date(2019, time.March, 25, 0, 0, 0, 0, time.UTC),
+				End:   time.Date(2019, time.March, 25, 0, 0, 1, 0, time.UTC),
+			},
+		},
+		{
+			desc: "year temporal",
+			notAfterInterval: &interval.Interval{
+				Start: time.Date(2019, time.March, 25, 0, 0, 0, 0, time.UTC),
+				End:   time.Date(2020, time.March, 25, 0, 0, 0, 0, time.UTC),
+			},
+		},
+	}
+
+	root, rootKey, err := rootAndKeySetup(rootFile, rootKeyFile)
+	if err != nil {
+		t.Fatalf("root and key setup error: %s", err)
+	}
+
+	timeNowUTC = func() time.Time {
+		return time.Date(2019, time.March, 25, 12, 0, 0, 0, time.UTC)
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			ca := &CA{
+				SigningCert: root,
+				SigningKey:  rootKey,
+				CertConfig: CertificateConfig{
```

Comment:
Seems that essentially this config remains fixed per the consts and only the notAfterInterval is varied.
In the interest of keeping the set-up code shorter and focusing on what varies from one test to another, how about:

(at the top of the file)
```
var leafTemplate = CertificateConfig{
  SubjectCommonName:         subjectCommonName,
  SubjectOrganization:       subjectOrganization,
  SubjectOrganizationalUnit: subjectOrganizationalUnit,
  SubjectLocality:           subjectLocality,
  SubjectCountry:            subjectCountry,
  SignatureAlgorithm:        signatureAlgorithm,
  DNSPrefix:                 prefix,
}
```
then making a local copy in the test body here:
```
  tmpl := leafTemplate
  tmpl.NotAfterInterval = test.notAfterInterval
  ca := &CA{ ... }
```